cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java

```diff
@@ -17,10 +17,170 @@
+                                                                               ProgramRunStatus.SUSPENDED);
+  private final Store store;
+  private final ProgramStateWriter programStateWriter;
+  private final ProgramRuntimeService runtimeService;
+  private final long startTimeoutSecs;
+  private final int txBatchSize;
+
+  RunRecordCorrectorService(CConfiguration cConf, Store store, ProgramStateWriter programStateWriter,
+                            ProgramRuntimeService runtimeService) {
+    this(store, programStateWriter, runtimeService, 2L * cConf.getLong(Constants.AppFabric.PROGRAM_MAX_START_SECONDS),
+         cConf.getInt(Constants.AppFabric.PROGRAM_RUNID_CORRECTOR_TX_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  RunRecordCorrectorService(Store store, ProgramStateWriter programStateWriter, ProgramRuntimeService runtimeService,
+                            long startTimeoutSecs, int txBatchSize) {
+    this.store = store;
+    this.programStateWriter = programStateWriter;
+    this.runtimeService = runtimeService;
+    this.startTimeoutSecs = startTimeoutSecs;
+    this.txBatchSize = txBatchSize;
+  }
+
+  void fixRunRecords() {
+    Set<ProgramRunId> fixed = doFixRunRecords();
+
+    if (!fixed.isEmpty()) {
+      LOG.info("Corrected {} of run records in {} status but no actual program running.",
+               NOT_STOPPED_STATUSES, fixed.size());
+    }
+  }
+
+  /**
+   * Fix all the possible inconsistent states for RunRecords that shows it is in RUNNING state but actually not
+   * via check to {@link ProgramRuntimeService} for a type of CDAP program.
+   *
+   * @return the set of fixed {@link ProgramRunId}.
+   */
+  private Set<ProgramRunId> doFixRunRecords() {
+    LOG.trace("Start getting run records not actually running ...");
+    // Get run records in STARTING, RUNNING and SUSPENDED states that are actually not running
+    // Do it in micro batches of transactions to avoid tx timeout
+
+    Set<ProgramRunId> fixedPrograms = new HashSet<>();
+    for (ProgramRunStatus status : NOT_STOPPED_STATUSES) {
+      long startTime = 0L;
+      while (true) {
+        Map<ProgramRunId, RunRecordMeta> runs = store.getRuns(status, startTime, Long.MAX_VALUE,
+                                                              txBatchSize, createFilter(fixedPrograms));
+        LOG.trace("{} run records in {} state but are not actually running", runs.size());
+        if (runs.isEmpty()) {
+          break;
+        }
+
+        for (RunRecordMeta record : runs.values()) {
+          startTime = Math.max(startTime, RunIds.getTime(record.getPid(), TimeUnit.SECONDS));
+
+          ProgramRunId programRunId = record.getProgramRunId();
+          if (!fixedPrograms.contains(programRunId)) {
+            programStateWriter.error(
+              programRunId,
+              new Throwable("Marking run record for " + programRunId + " as failed since no running program found."));
```

Comment:
Can we have a specific exception class for this? Seems odd to use Throwable...

Reply:
Sure.